### PR TITLE
refactor: refactor code and try to make it smaller

### PR DESCRIPTION
**Pull request by Uzaaft**
WIP attempt at making the code smaller.

### CODE REVIEW

**Comment by geohot** [comment]
The code in llvm and OpenCL isn't optimized and isn't well tested (probably shouldn't be touched), but I will merge the changes that remove the extra array construction from tuple, neaten things up, and move to f-strings.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -205,17 +205,17 @@ def matmul(self:Tensor, w:Tensor):
...
     if axis is None:
       axis = range(len(self.shape))
     if isinstance(axis, int):
       axis = [axis]
-    axis = tuple([x if x >= 0 else x+len(self.shape) for x in axis])
+    axis = tuple(x if x >= 0 else x+len(self.shape) for x in axis)
     shape = [self.shape[i] for i in range(len(self.shape)) if i not in axis]
     ret = fxn(self, axis=axis)
-    return ret if keepdim else ret.reshape(shape=[1] if shape == [] else shape)
+    return ret if keepdim else ret.reshape(shape=shape or [1])
```
Comment (geohot) [inline tinygrad/tensor.py:218]: Very cute

File: tinygrad/llops/ops_gpu.py [hunk 2]
```diff
@@ -146,12 +151,11 @@ def _ast(x: Union[GPUBuffer, LazyOp], buf_names: Dict[GPUBuffer, str], code_for_
...
         return "acc"
       srcs_code = [_ast(src, buf_names, code_for_op) for src in x.src]
-      code = code_for_op[x.op]
       if len(srcs_code) >= 1:
-        code = code.replace("A", srcs_code[0])
+        return code_for_op[x.op].replace("A", srcs_code[0])
       if len(srcs_code) >= 2:
-        code = code.replace("B", srcs_code[1])
-      return code
+        return code_for_op[x.op].replace("B", srcs_code[1])
```
Comment (geohot) [inline tinygrad/llops/ops_gpu.py:157]: Not right, you have to replace A also

Comment (Uzaaft) [reply tinygrad/llops/ops_gpu.py:157]: Ah I see. I misunderstood the control flow while refactoring. I'll fix it

**Comment by Uzaaft** [comment]
> The code in llvm and OpenCL isn't optimized and isn't well tested (probably shouldn't be touched), but I will merge the changes that remove the extra array construction from tuple, neaten things up, and move to f-strings.

Should I revert the changes done to the code in llvm and OpenCL?

**Comment by Uzaaft** [comment]
There is a lot of stuff in the `accel/` folder I'd like to try and clean up as fell. Would that contribute towards the overall line count? And is it of limits?

**Comment by geohot** [comment]
Yea, only in the core tinygrad repo, accel is all WIP.

**Comment by Uzaaft** [comment]
@geohot I've reverted the commits for the files I shouldn't touch.
Does everything look good for the core tinygrad repo?
I'll also try and reduce the lines in the `extra`, `tests` and `example` folders, since they count towards the total line count in the GitHub CI

**Comment by geohot** [comment]
You haven't, there's still changes in the accel/ folder.

**Comment by geohot** [comment]
Also in extra/, the issue is this stuff isn't well tested. The 1000 lines is only for the files under tinygrad/

**Comment by geohot** [comment]
Hmm perhaps these other changes are harmless enough. lmk when you are ready to merge.

**Comment by Uzaaft** [comment]
> Hmm perhaps these other changes are harmless enough. lmk when you are ready to merge.

I'm just fixing all the other files that I can touch xD Going trough the examples folder as we speak.

Ping me whenever the  code in llvm and OpenCL can be touched, and I'll happily take a look at those whenever they are ready. 

I'll ping you whenever I'm done with the last bit of the examples folder

**Comment by geohot** [comment]
Err, I wouldn't do examples either.

If you want to change code, you need a test for it.

**Comment by Uzaaft** [comment]
> If



> Err, I wouldn't do examples either.
> 
> If you want to change code, you need a test for it.

Even if it's just refactoring?

**Comment by Uzaaft** [comment]
@geohot Ready to review. 
Let me know when the code in `tinygrad/llops` can be touched. 
I see a few unneeded list comprehensions and string stuff there that can be moved to f-strings 👀

**Comment by geohot** [comment]
I don't think we're communicating well.

If you want to refactor code, you need coverage in a test. The code in tinygrad/ is covered, including tinygrad/llops, but not the llops that are symlinks from accel.

File: test/test_train.py [hunk 3]
```diff
@@ -15,9 +15,7 @@
 
 def train_one_step(model,X,Y):
   params = get_parameters(model)
-  pcount = 0
-  for p in params:
-    pcount += np.prod(p.shape)
+  pcount = sum(np.prod(p.shape) for p in params)
```
Comment (Pavocracy) [inline test/test_train.py:18]: I'm genuinely curious if there is a performance difference here, or is this purely just more compact?

(ps. I'm just here to learn about how to "neaten" up python code 😄)

Comment (Uzaaft) [reply test/test_train.py:18]: I did it mostly for the compactness.

But, generally, I think comprehensions are faster than for-loops, which this replaced.

Comment (Pavocracy) [reply test/test_train.py:18]: Thank you for replying. I might invest some effort in timing some basic comprehensions to replace my for loops to see the difference then :)

Comment (Uzaaft) [reply test/test_train.py:18]: Go for it.
My general rule:
If you are creating a new object, use comprehensions.
if you are doing a computation, you should stick with for-loops, or use array computations.

**Pull request closed without merging**